src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md

```diff
@@ -0,0 +1,345 @@
+
+The biggest hit is to large reads.  They drop from ~1 TiB/s to around 750 GiB/s.  Everything else sees a more modest, though consistent hit.  At this point I had to stop.  I really wanted to do PG scaling tests and even kernel RBD tests, but it was time to hand the systems back to the customer for re-imaging, followed by one of my excellent colleagues at Clyso to begin the integration process.
+
+# Finale
+
+So what's happened with this cluster since the end of the testing? All hardware was re-imaged and the new OSDs were deployed into the customer's existing HDD cluster.  Dan's upmap-remapped script is being used to control the migration process and we've migrated around 80% of the existing data to the NVMe backed OSDs.  By next week, the cluster should be fully migrated to the new NVMe based nodes.  We've opted not to employ all of the tuning we've done here, at least not at first.  Initially we'll make sure the cluster behaves well under the existing, mostly default configuration.  We now have a mountain of data we can use to tune the system further if the customer hits any performance issues.
+
+Since there was a ton of data and a ton of graphs here, I want to recap some of the highlights.  Here's an outline of the best numbers we were able to achieve on this cluster:
+
+|                | 30 OSDs (3x) | 100 OSDs (3x) | 320 OSDs (3x) | 630 OSDs (3x) | 630 OSDs (EC62)
+| -------------- | ------------ | ------------- | ------------- | ------------- | ---------------
+| Co-Located Fio | No           | No            | No            | Yes           | Yes
+| 4MB Read       | 63 GiB/s     | 214 GiB/s     | 635 GiB/s     | 1025 GiB/s    | 547 GiB/s
+| 4MB Write      | 15 GiB/s     | 46 GiB/s      | 133 GiB/s     | 270 GiB/s     | 387 GiB/s
+| 4KB Rand Read  | 1.9M IOPS    | 5.8M IOPS     | 16.6M IOPS    | 25.5M IOPS    | 3.4M IOPS
+| 4KB Rand Write | 248K IOPS    | 745K IOPS     | 2.4M IOPS     | 4.9M IOPS     | 936K IOPS
+
+What's next? We need to figure out how to fix the laggy PG issue during writes.  We can't have Ceph falling apart when the write workload gets too big.  Beyond that, we learned through this exercise that Ceph is perfectly capable of saturating 2x100GbE NICs.  To push the throughput envelope further we will need 200GbE+ when using 10 NVMe drives per node or more.  IOPS is more nuanced. We know that PG count can have a big effect.  We also know that the general OSD threading model is playing a big role.  We consistently hit a wall at around 400-600K random read IOPS per node and we've seen it in multiple deployments.  Part of this may be how the async msgr interfaces with the kernel and part of this may be how OSD threads wake up when new work is put into the shard queues.  I've modified the OSD code in the past to achieve better results under heavy load, but at the expense of low-load latency.  Ultimately, I suspect improving IOPS will take a multi-pronged approach and a rewrite of some of the OSD threading code.
```

Comment:
s/gets too big/scales up/
space after 2x -- suggest adding up at the top whether there was a separate physical or VLAN replication network, and if bonding was used, and with which hash policy

Reply:
Done for the first part.  For the second part, I'm going to leave that out for now.  If the customer wants to expose more details regarding their network configuration we can go back and add it as an addendum.  I just want to be a little cautious about sharing details of things we don't have explicit permission to share.